02-peer-protocol.md

```diff
@@ -49,11 +68,502 @@ transaction is confirmed are also not persistent - until you know the script
+
+The sending node:
+  - MUST add all sent outputs to the transaction
+  - if is the *initiator*:
+    - MUST send even `serial_id`s
+  - if is the *non-initiator*:
+    - MUST send odd `serial_id`s
+
+The receiving node:
+  - MUST add all received outputs to the transaction
+  - MUST accept P2WSH, P2WPKH, P2TR `script`s
+  - MAY fail the negotiation if `script` is non-standard
+  - MUST fail the negotiation if:
+    - the `serial_id` is already included in the transaction
+    - the `serial_id` has the wrong parity
+    - it has received 4096 `tx_add_output` messages during this negotiation
+    - the `sats` amount is less than the `dust_limit`
+    - the `sats` amount is greater than 2,100,000,000,000,000 (`MAX_MONEY`)
+
+#### Rationale
+
+Each node must know the set of the transaction outputs.
+
+`serial_id` is a randomly chosen number which uniquely identifies this output.
+Outputs in the constructed transaction are sorted by `serial_id`.
+
+`sats` is the satoshi value of the output.
+
+`script` is the scriptPubKey for the output (with its length omitted).
+`script`s are not required to follow standardness rules: non-standard
+scripts such as `OP_RETURN` may be accepted, but the corresponding
+transaction may fail to relay across the network.
+
+### The `tx_remove_input` and `tx_remove_output` Messages
+
+This message removes an input from the transaction.
+
+1. type: 68 (`tx_remove_input`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+
+This message removes an output from the transaction.
+
+1. type: 69 (`tx_remove_output`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+
+#### Requirements
+
+The sending node:
+  - MUST NOT send a `tx_remove` with a `serial_id` it did not add
+    to the transaction or has already been removed
+
+The receiving node:
+  - MUST remove the indicated input or output from the transaction
+  - MUST fail the negotiation if:
+    - the input or output identified by the `serial_id` was not added by the
+      sender
+    - the `serial_id` does not correspond to a currently added input (or output)
+
+### The `tx_complete` Message
+
+This message signals the conclusion of a peer's transaction
+contributions.
+
+1. type: 70 (`tx_complete`)
+2. data:
+    * [`channel_id`:`channel_id`]
+
+#### Requirements
+
+The nodes:
+  - MUST send this message in succession to conclude this protocol
+
+The receiving node:
+  - MUST use the negotiated inputs and outputs to construct a transaction
+  - MUST fail the negotiation if:
+    - the peer's total input satoshis is less than their outputs
+    - the peer's paid feerate does not meet or exceed the agreed `feerate`
+      (based on the `minimum fee`).
+    - if is the *non-initiator*:
+      - the *initiator*'s fees do not cover the `common` fields
+    - there are more than 252 inputs
+    - there are more than 252 outputs
+    - the estimated weight of the tx is greater than 400,000 (`MAX_STANDARD_TX_WEIGHT`)
```

Comment:
> I think the check should be delayed when all `tx_signatures` have been received ?

Unfortunately you can't, it would be too late by that time. If you're sending `tx_signatures` first, your peer can always choose to publish the transaction, even if their `tx_signatures` has a larger witness than expected. Failing at that point wouldn't be very interesting, you should either wait for the transaction to confirm or double-spend it if you don't see it confirming quickly enough (so we don't need to specifically handle this case).

Reply:
> If you're sending tx_signatures first, your peer can always choose to publish the transaction, even if their tx_signatures has a larger witness than expected. 

You could sequentialize the `tx_signatures` exchange as this witness weight estimation check is always a possibility for the first-receiver of the `tx_signatures`, though not for the remaining counterparties. I think this has been discussed with previous iterations of the spec, and it sounds the current proposal is “okayish” for now.

I still think it will be a source of DoS for collaborative transaction participants in the future and this stay an interesting problem. You might have margin of improvements like knowing the “worst-policy-size” and based on this ask premium fees for releasing signatures first. Though definitely beyond the scope of this PR.